### ISSUE TITLE
tm_inv

### ISSUE THREAD

**Issue by zvezdochiot** [mikhel1984/tmatrix#2]
Hi @mikhel1984 .

The standard error of most matrix libraries - absence of the condition number of the matrix `cond(A)`:  
![cond(A)](https://dxdt.ru/f/?cond(A_{n\times{n}})={{||A_{n\times{n}}||_2\times||A_{n\times{n}}^{-1}||_2}\over{n}})
```c
/* N = norm2(m) , 0.0 == error */
tmVal tm_norm2(tMat* m, int* err) 
{
  int i, j, R, C, e = 0;
  tmVal a, norm2 = 0.0;

  TM_ASSERT_ARGS(m, e, end_norm2);
   
  R = m->rows;
  C = m->cols; 
  for(i = 0; i < R; i++)
  {
    for(j = 0; j < C; j++) 
    {
      a = *tm_at(m,i,j);
      norm2 += (a * a);
    }
  }
  if (norm2 > 0.0)
    norm2 = sqrt(norm2);
  else
    e = TM_ERR_NO_SOLUTN;
  
end_norm2:
  if(err) *err = e;
  return norm2;
}

/* Ic = cond(m, minv) , 0.0 == error */
tmVal tm_cond(tMat *m, tMat *minv, int *err)
{
  int e = 0;
  tmVal cond = 0.0;
  int rA = tm_rank(m, &err);
  if (rA > 0.0)
    cond = tm_norm2(m, &err) * tm_norm2(minv, &err) / rA)
  else
    e = TM_ERR_NO_SOLUTN;
  if(err) *err = e;
  return cond;
}
```
so
```c
/* Inversion , 0.0 == error */
tmVal tm_inv(tMat *dst, tMat *m, int *err)
{
  tmVal cond = 0.0;
  ...
  cond = tm_cond(m, dst, &err);

end_inv:
  if(n > MEM_TMP_VEC) free(idx);
  tm_clear(&tmp);

  if (!(cond > 0.0))
    e = TM_ERR_NO_SOLUTN;
   
  if(err) *err = e;
  return cond;
}
```
and
```c
  tmVal condA = tm_inv(dst, m, &err);
  if ((condA > 0.0) && (condA < 1e14))
  {
    // good
    fpfintf(stdout, "cond(A)=%g\n", condA);
    ...
  }
  else
  {
    // error
    fpfintf(stderr, "ERROR: cond(A)=%g\n", condA);
    ...
  }
```
See also:
* [Wikipedia : Condition number](https://en.wikipedia.org/wiki/Condition_number)
* [Linear Algebra Operators and Generic Functions : cond](https://www.omnicalculator.com/math/condition-number)
* [MATLAB Function Reference : cond](http://matlab.izmiran.ru/help/techdoc/ref/cond.html)
* [SciLab : cond](https://help.scilab.org/docs/6.1.0/en_US/cond.html)
* [GNU Octave : Basic Matrix Functions](https://docs.octave.org/v4.4.0/Basic-Matrix-Functions.html)

**Comment by mikhel1984** [comment]
Hi @zvezdochiot! Thank you for this issue. I have added your suggestion with small modifications.

**Comment by zvezdochiot** [comment]
Hi @mikhel1984 .

To automatically close the issue (and leave a link to it), the commit is called **"fix #2: Add condition number checking"**.

See also **"Merge pull request..."** in history.
